### PR TITLE
Group shares with same source and target

### LINKED ISSUE #24575: Doubled folders when receiving shares through two groups
### Steps to reproduce
1. Create two usergroups G1 and G2
2. Share a folder F with both groups
3. Create a user U belonging to both groups G1 and G2
4. Log in with U

### Expected behaviour
User U will have folder F in his files.

### Actual behaviour
User U has folder F two times in his files (one is named "F (2)").

### Server configuration
**Operating system**: Debian 8.0

**Web server:** Apache 2.0

**Database:** MySQL

**PHP version:** 5.6.20

**ownCloud version:** 9.0.2

**Updated from an older ownCloud or fresh install:** fresh install

**Where did you install ownCloud from:** via official aptitude repo

**Signing status (ownCloud 9.0 and above):** No errors have been found.

**List of activated apps:**
```

  - activity: 2.2.1
  - activitydefaults: 0.1.0
  - comments: 0.2
  - dav: 0.1.6
  - direct_menu: 0.7.1
  - external: 1.2
  - federatedfilesharing: 0.1.0
  - federation: 0.0.4
  - files: 1.4.4
  - files_pdfviewer: 0.8.1
  - files_sharing: 0.9.1
  - files_texteditor: 2.1
  - files_trashbin: 0.8.0
  - files_versions: 1.2.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - gallery: 14.5.0
  - notifications: 0.2.3
  - ownbackup: 0.3.8
  - provisioning_api: 0.4.1
  - systemtags: 0.2
  - templateeditor: 0.1
  - updatenotification: 0.1.0
```

**The content of config/config.php:**
```
<?php
$CONFIG = array (
  'updatechecker' => false,
  'instanceid' => 'qwerty',
  'passwordsalt' => '0815',
  'secret' => 'thisissecret',
  'trusted_domains' =>
  array (
    0 => 'www.xxx.de',
    1 => 'xxx.de',
  ),
  'datadirectory' => '/var/www/vhosts/xxx.de/httpdocs/data',
  'overwrite.cli.url' => 'https://www.xxx.de',
  'dbtype' => 'mysql',
  'version' => '9.0.2.2',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'owncloud',
  'dbpassword' => 'dbpass',
  'logtimezone' => 'UTC',
  'installed' => true,
  'memcache.local' => '\\OC\\Memcache\\APCu',
  'mail_from_address' => 'info',
  'mail_smtpmode' => 'php',
  'mail_domain' => 'xxx.de',
  'theme' => 'mytheme',
);
```

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #24575

Note that this is a very limited solution and eventually we want smarter
merging!

TODO:
- [x] Add intergration tests
- [x] Check if permissions are properly accumulated
- [x] Check if javascript needs to be fixed

CC: @PVince81 @jonasheinisch

### CODE REVIEW

**Comment by rullzer** [comment]
Mmmm seems the js side needs more fixing...

**Comment by PVince81** [comment]
Wasn't there such logic already in the past ? Or did it get discarded through refactorings ?

**Comment by PVince81** [comment]
@icewind1991 do you remember anything ?

**Comment by rullzer** [comment]
Yes there was such logic. But it had to be discarded for various reasons.

* We can't group shares in the providers as there is no (future) guarantee that the current provider division is true for all installations
* We can't group shares in the providers since we support pagination. So if you only request the first N shares what to do.

At the same time I think this logic is now in the place where it should be. In the sharedMount/sharedStorage. Since that is the actual place where we group stuff.

**Comment by PVince81** [comment]
Doesn't help with https://github.com/owncloud/core/issues/25186 which will need a different fix.

**Comment by PVince81** [comment]
I don't know what to do with this. It feels a bit dangerous to merge that late.
Also, unit tests are missing.

File: core/js/shareitemmodel.js [hunk 1]
```diff
@@ -607,13 +607,17 @@
...
 				if (data2[0].ocs.data.length) {
 					reshare = data2[0].ocs.data[0];
+					reshares = data2[0].ocs.data;
 				}
 
 				model.set(model.parse({
 					shares: sharesMap,
-					reshare: reshare
+					reshare: reshare,
+					reshares: reshares,
```
Comment (PVince81) [inline core/js/shareitemmodel.js]: don't add extra commas in JS, this used to break in IE

File: apps/files_sharing/lib/SharedMount.php [hunk 2]
```diff
@@ -106,7 +112,11 @@ private function verifyMountPoint(\OCP\Share\IShare $share, array $mountpoints)
 	 */
 	private function updateFileTarget($newPath, &$share) {
 		$share->setTarget($newPath);
-		\OC::$server->getShareManager()->moveShare($share, $this->user);
+
+		foreach ($this->groupedShares as $share) {
```
Comment (PVince81) [inline apps/files_sharing/lib/SharedMount.php]: conflict with the "share" variable from the method argument.

Also I'm confused why we are relying on the internal `groupedShares` here and ignoring the `$share` attribute, the function doesn't fulfil its contract

Comment (rullzer) [inline apps/files_sharing/lib/SharedMount.php]: Ah fair enough.

Yeah it needs to be cleaned up more. Basically the groupedShares are always there now. The 'regual' share is just a super set of all shares. I did it this way so we have less code paths. So we even if you have just 1 incomming share then you will have a 'super' share with the same stuff. And `groupedShares` with just 1 entry.

I prefer 1 code path.

**Comment by PVince81** [comment]
Rebased

**Comment by PVince81** [comment]
Ouch, this breaks sharing: when opening the share panel it always says "Resharing not allowed" even for non-shared folders where I'm the owner.

**Comment by PVince81** [comment]
Some test cases for later, where "user1" is in "group1" and "group2"
- [ ] TEST: share folder "test" as outsider to "group1" and "user1" with same permissions
- [ ] TEST: share folder "test" as outsider to "group1" and "user1" with different permissions
- [ ] TEST: share folder "test" as "user1" to "group1"
- [ ] TEST: share folder "test" as outsider to "group1" and "group2" with same permissions
- [ ] TEST: share folder "test" as outsider to "group1" and "group2" with different permissions
- [ ] TEST: share folder "test" as outsider to "group1" and "group2" and "user1" (why not!) with different permissions
- [ ] TEST: share folder "test" as "user1" to "group1" and "group2" with same permissions
- [ ] TEST: share folder "test" as "user1" to "group1" and "group2" with different permissions

Most of these gave me two folders on v9.0.3 so let's hope this PR covers them all :smile:

**Comment by PVince81** [comment]
After reverting the JS commit locally I'm able to share again, so something to look into.

I did a quick test with the different grouping cases and it seems to work nicely ! Great job !

The hardest part: write integration tests for them.

**Comment by PVince81** [comment]
@rullzer any reason for the JS fix ? Without the JS there is no duplication in the "Shared with others" and "Shared with you" entries. Unless I missed a case ?

**Comment by PVince81** [comment]
ah, reshares.. forgot those :frowning:

**Comment by PVince81** [comment]
- [ ] TODO: need repair step to recombine the wrongly duplicated folders (even if they were manually excluded as a workaround)

Sounds like fun.

**Comment by PVince81** [comment]
- [ ] TEST: unshare from self for a grouped entry, which one is affected ?
- [ ] TEST: reshare grouped entry

**Comment by rullzer** [comment]
> > TODO: need repair step to recombine the wrongly duplicated folders (even if they were manually excluded as a workaround)

> Sounds like fun.

It is actually not that hard.
The fun part is deciding if you want ot merge all? Or just the ones still marked as dups ( (2), (3) ) etc.

**Comment by PVince81** [comment]
@rullzer ideally repair it to make it work just like it did in 8.2.6.
So all non de-duplicated entries are there for such cases, repair them to be merged properly.

**Comment by PVince81** [comment]
Added integration tests to match the cases from https://github.com/owncloud/core/pull/25113#issuecomment-231122523

@SergioBertolinSG check this out, I used some regexp magic to make some sentences more flexible

**Comment by PVince81** [comment]
Oh, looks like I had an outdated version of the branch. The tests above actually pass and the integration tests will prove it.

**Comment by PVince81** [comment]
- [ ] TODO: add unit tests for the improved MountProvider logic

File: apps/files_sharing/lib/SharedMount.php [hunk 3]
```diff
@@ -61,10 +64,13 @@ class SharedMount extends MountPoint implements MoveableMount {
 	public function __construct($storage, array $mountpoints, $arguments = null, $loader = null) {
 		$this->user = $arguments['user'];
 		$this->recipientView = new View('/' . $this->user . '/files');
-		$this->share = $arguments['newShare'];
-		$newMountPoint = $this->verifyMountPoint($this->share, $mountpoints);
+
+		$this->superShare = $arguments['superShare'];
+		$this->groupedShares = $arguments['groupedShares'];
```
Comment (PVince81) [inline apps/files_sharing/lib/SharedMount.php]: I suppose another enhanced approach would be to define a new class `CompositeShare` that implements `IShare` and contains all the children (groupedShares) then when we call `setTarget()` on the main one it internally sets the target on all sub-shares.

This could have the advantage of being reusable in other contexts where grouping is required.

File: apps/files_sharing/lib/MountProvider.php [hunk 4]
```diff
@@ -95,4 +100,63 @@ public function getMountsForUser(IUser $user, IStorageFactory $storageFactory) {
...
+	 * @return \OCP\Share\IShare[]
+	 */
+	private function groupShares(array $shares) {
+		$tmp = [];
+
+		foreach ($shares as $share) {
+			if (!isset($tmp[$share->getNodeId()])) {
+				$tmp[$share->getNodeId()] = [];
+			}
+			$tmp[$share->getNodeId()][$share->getTarget()][] = $share;
```
Comment (PVince81) [inline apps/files_sharing/lib/MountProvider.php]: @rullzer Shouldn't nodeId and target be combined into a single key ? That is, if you're grouping by these too, unless this is doing something slightly different.

Once I understand this I'll update the PHPDoc.

Comment (PVince81) [inline apps/files_sharing/lib/MountProvider.php]: Seems it would likely yield the same results anyway.

Whenever a single folder is has two shares, one with group and one with user, as long as the target is the same they need to stay combined. Once the user renamed the target, they need to stay split. That makes sense.

File: apps/files_sharing/lib/MountProvider.php [hunk 5]
```diff
@@ -95,4 +100,63 @@ public function getMountsForUser(IUser $user, IStorageFactory $storageFactory) {
 		// array_filter removes the null values from the array
 		return array_filter($mounts);
 	}
+
+	/**
+	 * @param \OCP\Share\IShare[] $shares
+	 * @return \OCP\Share\IShare[]
```
Comment (PVince81) [inline apps/files_sharing/lib/MountProvider.php]: so basically this is actually `\OCP\Share\IShare[][]`, first level is group and second is list of shares within that group

**Comment by PVince81** [comment]
@rullzer improved readability a bit, let me know whether this makes sense: https://github.com/owncloud/core/pull/25113/commits/4c46b4a16a1510e707b734d26f59e0fde20608e1

More unit tests to come later for better coverage of the different cases.

File: apps/files_sharing/lib/SharedMount.php (hunk 3 above)
Comment (rullzer) [inline apps/files_sharing/lib/SharedMount.php]: I like that!
very elegant and abstract nicely away

File: apps/files_sharing/lib/MountProvider.php (hunk 4 above)
Comment (rullzer) [inline apps/files_sharing/lib/MountProvider.php]: Yeah that is the 'old' behaviour anyways.

I'm not convinced that is always the right approach. Ideally I just see 1 entry per share. Else you also get duplicated data if you sync etc.

But lets make it behave like in the old days first.

**Comment by rullzer** [comment]
@PVince81 had a quick look. Looks good. Will have a deeper look after some sleep.

**Comment by rullzer** [comment]
@PVince81 after a night sleep. You fix indeed cleans it up! Thanks.

**Comment by PVince81** [comment]
Working on the repair steps now.

I find the hardest part is to identify which to adjust and which to leave alone.

This is on 9.0: when share with group1, group2 and user1 (user1 is in group1 and group2)
```
MariaDB [owncloud]> select * from oc_share order by item_source, uid_owner, stime;
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+---------------+-------------+------------+----------+------------+-------+-----------+
| id | share_type | share_with | uid_owner | uid_initiator | parent | item_type | item_source | item_target | file_source | file_target   | permissions | stime      | accepted | expiration | token | mail_send |
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+---------------+-------------+------------+----------+------------+-------+-----------+
|  1 |          1 | group1     | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test         |          31 | 1468329809 |        0 | NULL       | NULL  |         0 |
|  6 |          2 | user1      | admin     | admin         |      1 | folder    | 9           | NULL        |           9 | /test (2)     |          31 | 1468329809 |        0 | NULL       | NULL  |         0 |
|  2 |          1 | group2     | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test         |          31 | 1468329812 |        0 | NULL       | NULL  |         0 |
|  3 |          2 | user1      | admin     | admin         |      2 | folder    | 9           | NULL        |           9 | /test (2) (2) |          31 | 1468329812 |        0 | NULL       | NULL  |         0 |
|  5 |          0 | user1      | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test         |          31 | 1468329916 |        0 | NULL       | NULL  |         0 |
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+---------------+-------------+------------+----------+------------+-------+-----------+
```

And here is when created on this branch:
```
MariaDB [owncloud]> select * from oc_share order by item_source, uid_owner, stime;
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+-------------+-------------+------------+----------+------------+-------+-----------+
| id | share_type | share_with | uid_owner | uid_initiator | parent | item_type | item_source | item_target | file_source | file_target | permissions | stime      | accepted | expiration | token | mail_send |
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+-------------+-------------+------------+----------+------------+-------+-----------+
|  7 |          1 | group1     | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test       |          31 | 1468330088 |        0 | NULL       | NULL  |         0 |
|  8 |          1 | group2     | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test       |          31 | 1468330091 |        0 | NULL       | NULL  |         0 |
|  9 |          0 | user1      | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test       |          31 | 1468330095 |        0 | NULL       | NULL  |         0 |
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+-------------+-------------+------------+----------+------------+-------+-----------+
```

However, it could happen that the user decided to rename (also on this branch):
```
MariaDB [owncloud]> select * from oc_share order by item_source, uid_owner, stime;
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+---------------+-------------+------------+----------+------------+-------+-----------+
| id | share_type | share_with | uid_owner | uid_initiator | parent | item_type | item_source | item_target | file_source | file_target   | permissions | stime      | accepted | expiration | token | mail_send |
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+---------------+-------------+------------+----------+------------+-------+-----------+
|  7 |          1 | group1     | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test         |          31 | 1468330088 |        0 | NULL       | NULL  |         0 |
| 10 |          2 | user1      | admin     | admin         |      7 | folder    | 9           | NULL        |           9 | /test renamed |          31 | 1468330088 |        0 | NULL       | NULL  |         0 |
|  8 |          1 | group2     | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test         |          31 | 1468330091 |        0 | NULL       | NULL  |         0 |
| 11 |          2 | user1      | admin     | admin         |      8 | folder    | 9           | NULL        |           9 | /test renamed |          31 | 1468330091 |        0 | NULL       | NULL  |         0 |
|  9 |          0 | user1      | admin     | admin         |   NULL | folder    | 9           | NULL        |           9 | /test renamed |          31 | 1468330095 |        0 | NULL       | NULL  |         0 |
+----+------------+------------+-----------+---------------+--------+-----------+-------------+-------------+-------------+---------------+-------------+------------+----------+------------+-------+-----------+
```

This looks very close to the bogus case that needs to be repaired.

So for now the idea is to first gather all shares with type "2" (children of group shares) and check whether they all have the same "target". If they do, it means it was a legitimate rename.
However, if the target is inconsistent, then set the target to the one of the parent.

**Comment by PVince81** [comment]
and take into account the case where the user voluntarily deleted the double shares in 9.0 and restore them by setting their permissions back for proper grouping.

**Comment by PVince81** [comment]
- [ ] TODO: add some integration test cases for when the recipient renamed the target, make sure that the old target and no duplicates appear

**Comment by PVince81** [comment]
Okay, turns out that a voluntary rename will cause one child entry for **every** group share.
While a bogus entry will only have a single child entry with a different name, attached to only one of the groups.

Bogus:
```
MariaDB [owncloud]> select id,uid_owner,item_source,share_type,parent,file_target,permissions from oc_share order by item_source, uid_owner, stime;
+----+-----------+-------------+------------+--------+-------------+-------------+
| id | uid_owner | item_source | share_type | parent | file_target | permissions |
+----+-----------+-------------+------------+--------+-------------+-------------+
| 12 | admin     | 9           |          1 |   NULL | /test       |          31 |
| 13 | admin     | 9           |          1 |   NULL | /test       |          31 |
| 14 | admin     | 9           |          2 |     13 | /test (2)   |          31 |
+----+-----------+-------------+------------+--------+-------------+-------------+
```

Legitimate rename:
```
MariaDB [owncloud]> select id,uid_owner,item_source,share_type,parent,file_target,permissions from oc_share order by item_source, uid_owner, stime;
+----+-----------+-------------+------------+--------+-------------+-------------+
| id | uid_owner | item_source | share_type | parent | file_target | permissions |
+----+-----------+-------------+------------+--------+-------------+-------------+
| 15 | admin     | 9           |          1 |   NULL | /test       |          31 |
| 17 | admin     | 9           |          2 |     15 | /test (2)   |          31 |
| 16 | admin     | 9           |          1 |   NULL | /test       |          31 |
| 18 | admin     | 9           |          2 |     16 | /test (2)   |          31 |
+----+-----------+-------------+------------+--------+-------------+-------------+
```

**Comment by PVince81** [comment]
@rullzer here's a beginning a made before the realization in my former comment.

Did you have any fancy/clever ideas since it sounded like it would be easy ?
https://github.com/owncloud/core/pull/25113/commits/deca93306995abbdf03ff4be125009592f53fe08

Maybe you were thinking of another approach like iterating over every user and then grouping for them using the same logic as in MountProviderTest ?

File: lib/private/Repair/RepairUnmergedShares.php [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		$maxPermissions = 0;
+		$mainShare = null;
+
+		do {
+			$resultsCount = 0;
+			$query->setFirstResult($offset);
+			$resultSet = $query->execute();
+
+			while (($result = $resultSet->fetch()) !== false) {
+				$key = $result['item_source'] . '/' . $result['uid_owner'];
```
Comment (rullzer) [inline lib/private/Repair/RepairUnmergedShares.php]: Each item only has 1 owner right?

**Comment by PVince81** [comment]
Looks like we'll also need to find group memberships.

Maybe my approach is wrong and we should iterate over users instead of over shares.

File: lib/private/Repair/RepairUnmergedShares.php (hunk 6 above)
Comment (PVince81) [inline lib/private/Repair/RepairUnmergedShares.php]: indeed, just wanted to be safe in case we have older unfixed reshares where the owner might be the sharer instead of the original owner

**Comment by PVince81** [comment]
- [ ] TODO: also need some testing with subfolder reshares (shares within a received parent folder)

**Comment by PVince81** [comment]
- [ ] TODO: exclude repair step if share provider is not the default one ?

**Comment by PVince81** [comment]
@rullzer I completely revamped the repair step, please check it out.
That's quite some complex machinery.

Now it does the following

1. Iterate over every user
1. Get the subshares (SHARE_TYPE_USERGROUP)
1. Get the group shares (SHARE_TYPE_GROUP)
1. Get the user shares (SHARE_TYPE_USER)
1. Group them by "item_source"
1. Iterate over "item_source"
1. For each set of shares (we consider all three types to be a single share entry)
1. Check if the share is valid:
    - a valid share has one subshare that matches every group share
    - a valid share has the same "file_target" value in every subshare
    - the "file_target" can be different than the one from the group share (legitimate reshare) as long as every subshare has the same
1. If the share is not valid, fix it
    - set the "file_target" to the same value as the first group share
    - if "permissions" is 0 it means the user might have deleted some of the duplicates, restore these to the permissions from the parent group share

Note that the "SHARE_WITH_USER" entry is treated like a subshare and also gets its "file_target" adjust.

Now I realize I missed something:
- [ ] TODO: if the user opted out of ALL subshares (deleted all duplicates + the main share), then only adjust "file_target"

**Comment by PVince81** [comment]
I tested subfolder reshares and they are not affected. It doesn't cause any duplicate because the item inside "test/inside" and "test (2)/inside" is technically the same folder, same fileid, so any reshare created there will be in the correct format.

**Comment by PVince81** [comment]
Okay, repair step adjusted. I think I'm done here and all known cases should be adressed.
@rullzer would be good if you could have a look. See https://github.com/owncloud/core/pull/25113#issuecomment-232113063 for the repair step summary. Maybe I could add this in the class description.

Next step: the JS part.

**Comment by PVince81** [comment]
Added more unit tests.

Now all that is left is the JS part.

**Comment by PVince81** [comment]
- [ ] TODO: tweak version check for the repair step. Some people might be on 9.1.0RC1 but also need repair

**Comment by PVince81** [comment]
And JS part done.

@owncloud/sharing please review and test.

Please note that I'll have to adjust the version check here https://github.com/owncloud/core/pull/25113/files#diff-638fff1908dfeea3094e9060752f9b0fR305

For the merge on 9.2/master and 9.1.1/stable9.1, it will be "run repair step if coming from OC < 9.1.1".
For the backport on 9.0 it will be "run the repair step if coming from OC < 9.0.4".

**Comment by PVince81** [comment]
A To test the repair step

1. Setup OC <= 9.0.3
1. Create shares based on the scenarios from https://github.com/owncloud/core/pull/25113#issuecomment-231122523
1. Check that the recipient has duplicated received folders (you might delete some, for testing)
1. Run upgrade to this branch
1. Check that the recipient has no more duplicates

B To test creating shares

1. Setup OC on this branc
1. Create shares based on the scenarios from https://github.com/owncloud/core/pull/25113#issuecomment-231122523
1. Check that the recipient has no duplicate received shares

@owncloud/qa

**Comment by SergioBertolinSG** [comment]
Integration tests pass 💚

**Comment by SergioBertolinSG** [comment]
After an upgrade from stable9, duplicated folders are still there, not cleaned. 
Sharing another folder in the same way didn't create duplicated folders.

**Comment by PVince81** [comment]
@SergioBertolinSG this could be because of the version check that doesn't work for all upgrade paths.

Try changing this line https://github.com/owncloud/core/pull/25435 to `if (true) {`.
Then run `maintenance:repair` (or the upgrade).

I still need to figure out a good way to run the repair only once. Maybe a flag in oc_appconfig ?

**Comment by SergioBertolinSG** [comment]
What line? it is missing in the link

**Comment by PVince81** [comment]
Clipboard fail, sorry...

Here's the link https://github.com/owncloud/core/pull/25113/files#diff-638fff1908dfeea3094e9060752f9b0fR305

**Comment by rullzer** [comment]
:+1:
Awesome work.

I still like the combinedshare idea. But later I guess.

**Comment by SergioBertolinSG** [comment]
Changing that line and running mainteinance:repair it has worked correctly.

**Comment by PVince81** [comment]
@SergioBertolinSG I have now adjusted the version check. If you upgrade from stable9 to this branch it should now properly run the repair.

When backporting this to 9.1 and 9.0 we'll need to adjust the versions again.

**Comment by SergioBertolinSG** [comment]
Works 👍

**Comment by PVince81** [comment]
Rebased for CI

**Comment by PVince81** [comment]
Rebased and adjusted version for 9.2

**Comment by PVince81** [comment]
how many timeswill I need to rebase this ?

**Comment by PVince81** [comment]
Close enough for CI => merge

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/issues/25534
stable9: TODO

**Comment by PVince81** [comment]
Looks like stable9 backport will be tricky, the old code is using old APIs and an old way. Or might require backporting more stuff.

**Comment by PVince81** [comment]
From what I see a lot of the code would require this commit https://github.com/owncloud/core/commit/6123badbfa22b79167c4b41f84a38b372f033a2e from https://github.com/owncloud/core/pull/23919.

Because on stable9 the MountProvider and SharedStorage still use the old array-style shares instead of the ShareManager style. I'm not sure I want to risk backporting this, so will see if there is a way to make it work with the old code.

**Comment by PVince81** [comment]
stable9 PR https://github.com/owncloud/core/pull/25543

**Comment by PVince81** [comment]
Looks like this fix is not enough, there's another scenario that is not covered: https://github.com/owncloud/core/issues/25564
This is getting tiring...
